pkg/apis/core/validation/validation.go

```diff
@@ -2546,6 +2546,9 @@ func validateProbe(probe *core.Probe, fldPath *field.Path) field.ErrorList {
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.PeriodSeconds), fldPath.Child("periodSeconds"))...)
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.SuccessThreshold), fldPath.Child("successThreshold"))...)
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.FailureThreshold), fldPath.Child("failureThreshold"))...)
+	if probe.TerminationGracePeriodSeconds != nil && *probe.TerminationGracePeriodSeconds <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("terminationGracePeriodSeconds"), *probe.TerminationGracePeriodSeconds, "must be greater than 0"))
+	}
```

Comment:
1.  According to https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/, “The value zero indicates stop immediately via the kill signal" , in other word, terminationGracePeriodSeconds 
can be set to zero. 
2. should we limit probe.TerminationGracePeriodSeconds to a max value?

Reply:
> “The value zero indicates stop immediately via the kill signal" , in other word, terminationGracePeriodSeconds
can be set to zero.

I don't think this was intended, which is why we are fixing this. As the feature author, I think this should never be set to non-negative, but I had written it to be consistent with the existing pod-level terminationGracePeriodSeconds.

See discussion around https://github.com/kubernetes/kubernetes/pull/98866#issuecomment-849259855 for more detail.

Reply:
> I think this should never be set to non-negative

that just did not never not break my brain :)

in the abstract, a probe asking to be killed at its timeout immediately by setting probe terminationGracePeriodSeconds=0 doesn't seem terrible to me... is there a problem with that I'm not seeing?

Reply:
Updated comment for clarity.

We don't want container specs setting terminationGracePeriodSeconds to 0 because that would cause a force deletion of the pod and unintentionally bypasses normal pod shutdown. Failed probes should not be able to do force deletions. The goal of this feature is to decouple the grace period on probe terminations from e.g. a user-initiated graceful termination, not add new unexpected behaviour :)